File: StarcraftNN/TrainingModule/BasicInterface.cpp
#include "BasicInterface.h"

BasicInterface::BasicInterface() : OrganismInterface() {
}

void BasicInterface::sendInputs(Organism* organism) {
  Network* network = organism->net;
  std::vector<float> sense_values;
  int sensorCount = _unitCount * 2;
  sense_values.resize(sensorCount);
  int sensor = 0;
  BWAPI::Position squadPosition = getCentroid(_allies);
  for(int i = 0; i < _allies.size(); i++) {
    BWAPI::Unit* unit = _allies[i];
    float distance = squadPosition.getDistance(unit->getPosition());
    sense_values[sensor++] = distance;
    BWAPI::Position difference = unit->getPosition() - squadPosition;
    float angle = tanf(difference.y() / difference.x());
    sense_values[sensor++] = angle;
  }
  network->load_sensors(sense_values);
  network->activate();
}

void BasicInterface::applyOutputs(Organism* organism) {
  std::vector<NNode*> outputs = organism->net->outputs;
  std::vector<ScoredUnit> _sunits;
  for(int i = 0; i < _unitCount; i++) {
    for(int j = 0; j < _unitCount; j++) {
      double attack = outputs[i * _unitCount + j]->activation;
      _sunits.push_back(ScoredUnit(_enemies[j], attack));
    }
    sort(_sunits.begin(), _sunits.end(), ScoredUnit::compare);
    
    BWAPI::Unit* ally = _allies[i];
    if(unitExists(ally)) {
      for(int j = 0; j < _sunits.size(); j++) {
        BWAPI::Unit* enemy = _sunits[j].unit;
        if(unitExists(enemy)) {
          if(_lastAttack[i] != enemy) {
            ally->attack(enemy);
            _lastAttack[i] = enemy;
          }
          break;
        }
      }
    }
  }
}

Genome* BasicInterface::createGenome() {
  Genome* g = new Genome(1, _unitCount * 2, _unitCount * _unitCount, 5, 100, true, .5);
  return g;
}

File: StarcraftNN/StarcraftNN/OrganismInterfaces/BasicInterface.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using SWIG.BWAPI;
using BwapiPosition = SWIG.BWAPI.Position;
using SharpNeat.Genomes.Neat;
using SharpNeat.Decoders.Neat;

namespace StarcraftNN.OrganismInterfaces
{
    public class BasicInterface : IOrganismInterface
    {
        private List<Unit> _allies, _enemies, _lastAttack;
        private static int UnitCount = 3;

        public string SaveFile
        {
            get
            {
                return "basic";
            }
        }

        private struct ScoredUnit
        {
            public Unit unit;
            public double score;
        }

        public BasicInterface()
        {
            _lastAttack = new List<Unit>();
            for (int i = 0; i < BasicInterface.UnitCount; i++)
                _lastAttack.Add(null);
        }

        public NeatGenomeFactory CreateGenomeFactory()
        {
            NeatGenomeFactory factory = new NeatGenomeFactory(UnitCount * 2, UnitCount * 3);
            return factory;
        }

        public void InputActivate(NeatGenome genome)
        {
            int sensor = 0;
            var allyPosition = Utils.getCentroid(_allies);
            NeatGenomeDecoder decoder = new NeatGenomeDecoder(SharpNeat.Decoders.NetworkActivationScheme.CreateCyclicFixedTimestepsScheme(1));
            var blackbox = decoder.Decode(genome);
            foreach (var enemy in _enemies)
            {
                BwapiPosition difference = enemy.getPosition().opSubtract(allyPosition);
                double distance = allyPosition.getDistance(enemy.getPosition());
                double angle = Math.Tan(difference.yConst() / difference.xConst());
                blackbox.InputSignalArray[sensor++] = distance;
                blackbox.InputSignalArray[sensor++] = angle;
            }
            blackbox.Activate();
            for (int i = 0; i < BasicInterface.UnitCount; i++)
            {
                Unit ally = _allies[i];
                List<ScoredUnit> scoredUnits = new List<ScoredUnit>();
                for (int j = 0; j < BasicInterface.UnitCount; j++)
                {
                    ScoredUnit su = new ScoredUnit();
                    su.score = blackbox.OutputSignalArray[i * BasicInterface.UnitCount + j];
                    su.unit = _enemies[j];
                    scoredUnits.Add(su);
                }
                scoredUnits.Sort((x, y) => { return y.score.CompareTo(x.score); });
                if (ally.exists())
                {
                    foreach (var enemy in _enemies)
                    {
                        if (enemy.exists())
                        {
                            if (_lastAttack[i] != enemy)
                            {
                                _lastAttack[i] = enemy;
                                ally.attack(enemy);
                            }
                            break;
                        }
                    }
                }
            }
        }

        public void UpdateState()
        {
            _allies = Utils.getAllies();
            _enemies = Utils.getEnemies();
        }

        public double ComputeFitness(int frameCount)
        {
            double maxScale = 4;
            double minimum = -(UnitCount * UnitCount) * maxScale;
            double score = Utils.getAllies().Count - Utils.getEnemies().Count;
            score *= Math.Abs(score);
            score *= 200.0 / frameCount;
            score -= minimum;
            return score;
        }
    }
}
